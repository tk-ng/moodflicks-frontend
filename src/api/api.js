import axios from "axios";

const baseURL = import.meta.env.VITE_BASE_URL || "http://localhost:3001/";

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

export class Api {
	// the token for interactive with the API will be stored here.
	static token;

	static async request(url, data = {}, method = "get") {
		// console.debug("API Call:", url, data, method);

		const headers = { Authorization: `Bearer ${Api.token}` };
		const params = method === "get" ? data : {};

		try {
			return (
				await axios({
					url,
					method,
					baseURL,
					data,
					params,
					headers,
					allowAbsoluteUrls: false,
				})
			).data;
		} catch (err) {
			console.error("API Error:", err.response);
			let message = err.response.data.error.message;
			throw Array.isArray(message) ? message : [message];
		}
	}

	// Individual API routes

	static async getUser(username) {
		let res = await this.request(`/users/${username}`);
		return res.user;
	}

	static async saveChanges({ username, data }) {
		let res = await this.request(
			`/users/${username}`,
			{ username, data },
			"patch"
		);
		return res.user;
	}

	static async deleteUser(username) {
		let res = await this.request(`/users/${username}`, {}, "delete");
		return res.user;
	}

	static async login(data) {
		let res = await this.request("auth/token", data, "post");
		return res.token;
	}

	static async signup(data) {
		let res = await this.request("auth/register", data, "post");
		return res.token;
	}

	static async getTrendingMovies() {
		let res = await this.request("/movies");
		return res.movies;
	}

	static async searchMovie(keyword) {
		let res = await this.request(`/movies/search/?keyword=${keyword}`);
		return res.movies;
	}

	static async getMovieDetail(id) {
		let res = await this.request(`/movies/${id}`);
		return res.movie;
	}

	static async getMovieCredits(id) {
		let res = await this.request(`/movies/${id}/credits`);
		return res.credits;
	}

	static async getActivities() {
		let res = await this.request(`/moods/activities`);
		return res.activities;
	}

	static async searchMood(keyword) {
		let res = await this.request(`/moods/search/?keyword=${keyword}`);
		return res.moods;
	}

	static async getMoods() {
		let res = await this.request(`/moods`);
		return res.moods;
	}

	static async getMoodsByUser(username) {
		let res = await this.request(`/users/${username}/moods`);
		return res.moods;
	}

	static async getMood(id) {
		let res = await this.request(`/moods/${id}`);
		return res.mood;
	}

	static async getMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/movies`);
		return res.movies;
	}

	static async getMoodsByMovie(id) {
		let res = await this.request(`/movies/${id}/moods`);
		return res.moods;
	}

	static async getSimilarMoods(id) {
		let res = await this.request(`/moods/${id}/similar`);
		return res.moods;
	}

	static async getRecommendedMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/recommended`);
		return res.movies;
	}

	static async createMood({ mood, username }) {
		let res = await this.request(`/moods`, { mood, username }, "post");
		return res.mood;
	}

	// needs username to verify in backend
	static async saveMoodChanges({ id, name, username }) {
		let res = await this.request(`/moods/${id}`, { name, username }, "patch");
		return res.message;
	}

	// needs username to verify in backend
	static async addMovieToMood({ moodId, movieId, username }) {
		const movies = await this.getMoviesByMood(moodId);

		if (movies.length) {
			const movieIds = movies.map((m) => m.id);
			if (movieIds.includes(movieId)) return null;
		}

		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"post"
		);
		return res.movieMood;
	}

	// needs username to verify in backend
	static async removeMovieFromMood({ moodId, movieId, username }) {
		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"delete"
		);
		return res.message;
	}

	// needs username to verify in backend
	static async deleteMood({ id, username }) {
		let res = await this.request(`/moods/${id}`, { username }, "delete");
		return res.message;
	}

	static async voteMood({ moodId, voteType, username }) {
		let res = await this.request(
			`/moods/${moodId}/vote/${voteType}`,
			{ username },
			"post"
		);
		return res.message;
	}
}
