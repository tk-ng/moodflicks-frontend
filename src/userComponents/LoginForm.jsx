import { useFields } from "../hooks/useFields";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { useNavigate } from "react-router";
import { toast } from "react-toastify";

export const LoginForm = ({ login }) => {
	const navigate = useNavigate();

	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await login(formData);
			resetForm();
			navigate("/");
			toast.success(`Welcome back, ${formData.username}!`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="LoginForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Log In</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
