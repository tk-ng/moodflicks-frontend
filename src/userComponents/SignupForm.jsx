import { useFields } from "../hooks/useFields";
import { useNavigate } from "react-router";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";

export const SignupForm = ({ signup }) => {
	const naviagte = useNavigate();
	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
			firstName: "",
			lastName: "",
			email: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await signup(formData);
			resetForm();
			naviagte("/");
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="SignupForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Sign Up</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
