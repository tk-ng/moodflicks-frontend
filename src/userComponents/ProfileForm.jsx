import { useFields } from "../hooks/useFields";
import { useNavigate, useParams, Navigate } from "react-router";
import { Api } from "../api/api";
import { UserContext } from "./UserContext";
import { useContext, useState, useEffect } from "react";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { toast } from "react-toastify";

export const ProfileForm = () => {
	const { currUser, setCurrentUser } = useContext(UserContext);

	const params = useParams();
	const naviagte = useNavigate();
	const [profileUser, setProfileUser] = useState();

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	const [
		formData,
		handleChange,
		resetForm,
		formErrors,
		setFormErrors,
		setFormData,
	] = useFields({
		firstName: "",
		lastName: "",
		email: "",
	});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			const updatedUser = await Api.saveChanges({
				username: profileUser.username,
				data: formData,
			});
			if (currUser.username === profileUser.username)
				setCurrentUser((currentUser) => ({
					...currentUser,
					data: updatedUser,
				}));
			resetForm();
			toast.success("Profile Updated Successfully");
			naviagte(`/users/${profileUser.username}`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	async function getUserInfo(username) {
		try {
			setIsLoading(true);
			setNotFound(false);

			const userResult = await Api.getUser(username);
			setProfileUser(userResult);
			setFormData({
				firstName: userResult.firstName,
				lastName: userResult.lastName,
				email: userResult.email,
			});
		} catch (err) {
			console.error("User not found:", err);
			setNotFound(true);
		} finally {
			setIsLoading(false);
		}
	}

	useEffect(() => {
		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);

		return <Navigate to="/" />;
	}

	if (!currUser.isAdmin && params.username !== currUser.username) {
		toast.warning(`You are not logged in as ${params.username}.`);
		return <Navigate to={`/users/${currUser.username}`} />;
	}

	return (
		<div className="ProfileForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Profile</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={params.username}
										disabled
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Save Changes</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
