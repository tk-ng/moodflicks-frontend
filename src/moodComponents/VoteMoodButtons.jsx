import { Button } from "reactstrap";
import { Api } from "../api/api";
import { UserContext } from "../userComponents/UserContext";
import { useContext } from "react";

export const VoteMoodButtons = ({ moodId, setMood }) => {
	const { currUser } = useContext(UserContext);

	const handleClick = async (voteType) => {
		await Api.voteMood({ moodId, voteType, username: currUser.username });
		setMood(await Api.getMood(moodId));
	};

	return (
		<div className="d-flex justify-content-around">
			<Button color="primary" onClick={() => handleClick("up")}>
				👍
			</Button>
			<Button color="danger" onClick={() => handleClick("down")}>
				👎
			</Button>
		</div>
	);
};
