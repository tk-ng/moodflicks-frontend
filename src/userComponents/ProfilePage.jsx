import { useContext, useEffect, useState } from "react";
import { UserContext } from "./UserContext";
import { Navigate, useParams } from "react-router";
import { Api } from "../api/api";
import { Button } from "reactstrap";
import { useNavigate } from "react-router";
import { MoodList } from "../moodComponents/MoodList";
import { toast } from "react-toastify";

export const ProfilePage = () => {
	const params = useParams();
	const navigate = useNavigate();
	const [user, setUser] = useState();
	const [moods, setMoods] = useState([]);
	const { currUser } = useContext(UserContext);

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	useEffect(() => {
		async function getUserInfo(username) {
			try {
				setIsLoading(true);
				setNotFound(false);
				const [userResult, moodsResult] = await Promise.all([
					Api.getUser(username),
					Api.getMoodsByUser(username),
				]);
				setUser(userResult);
				setMoods(moodsResult);
			} catch (err) {
				console.error("User not found:", err);
				setNotFound(true);
			} finally {
				setIsLoading(false);
			}
		}

		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);
		return <Navigate to="/" />;
	}

	return (
		<div className="text-light">
			<span className="display-6">{user.username}</span>
			{currUser && user.username === currUser.username && (
				<Button
					onClick={() => navigate(`/users/${user.username}/edit`)}
					className="m-3 "
				>
					Edit Profile
				</Button>
			)}
			<div>
				{moods.length ? (
					<>
						<h3> Mood-based movie lists, curated by {user.username}</h3>
						<MoodList moods={moods} hideCreatedBy={true} />
					</>
				) : (
					<span>{user.username} has not created any mood lists yet.</span>
				)}
			</div>
		</div>
	);
};
