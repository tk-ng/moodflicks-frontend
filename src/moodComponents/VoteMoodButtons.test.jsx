import { render, screen, waitFor } from "@testing-library/react";
import { VoteMoodButtons } from "./VoteMoodButtons";
import { UserContext } from "../userComponents/UserContext";
import { MemoryRouter } from "react-router";
import { Api } from "../api/api";
import userEvent from "@testing-library/user-event";
import { expect } from "vitest";

vi.mock("../api/api", () => ({
	Api: {
		voteMood: vi.fn(),
		getMood: vi.fn(),
		getMoods: vi.fn(),
		searchMood: vi.fn(),
	},
}));

describe("Render VoteMoodButtons component", function () {
	it("matches snapshot", function () {
		const { asFragment } = render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
					}}
				>
					<VoteMoodButtons />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it("mounts without crashing without moods", async function () {
		const { container } = render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
					}}
				>
					<VoteMoodButtons />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
	});
});

describe("Handle VoteMoodButtons onClick", function () {
	it("clicks the thumbs up button", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
					}}
				>
					<VoteMoodButtons moodId={"testMood"} setMood={vi.fn()} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		const user = userEvent.setup();
		await user.click(screen.getByText("üëç"));
		expect(Api.voteMood).toHaveBeenCalledWith({
			moodId: "testMood",
			voteType: "up",
			username: "testuser",
		});
		expect(Api.getMood).toHaveBeenCalledWith("testMood");
	});

	it("clicks the thumbs down button", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
					}}
				>
					<VoteMoodButtons moodId={"testMood"} setMood={vi.fn()} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		const user = userEvent.setup();
		await user.click(screen.getByText("üëé"));
		expect(Api.voteMood).toHaveBeenCalledWith({
			moodId: "testMood",
			voteType: "down",
			username: "testuser",
		});
		expect(Api.getMood).toHaveBeenCalledWith("testMood");
	});
});
