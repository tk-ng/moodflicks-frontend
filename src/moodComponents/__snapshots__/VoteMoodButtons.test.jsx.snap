// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render VoteMoodButtons component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-around"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      👍
    </button>
    <button
      class="btn btn-danger"
      type="button"
    >
      👎
    </button>
  </div>
</DocumentFragment>
`;
